File: java/secret-handshake/src/main/java/HandshakeCalculator.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class HandshakeCalculator {

    List<Signal> calculateHandshake(int number) {
        List<Signal> result = new ArrayList<>();
        if ((number & 1) == 1) result.add(Signal.WINK);
        if ((number & 2) == 2) result.add(Signal.DOUBLE_BLINK);
        if ((number & 4) == 4) result.add(Signal.CLOSE_YOUR_EYES);
        if ((number & 8) == 8) result.add(Signal.JUMP);
        if ((number & 16) == 16) Collections.reverse(result);
        return result;
    }

}

File: kotlin/secret-handshake/src/main/kotlin/HandshakeCalculator.kt
object HandshakeCalculator {
    fun calculateHandshake(i: Int): List<Signal> = mutableListOf<Signal>().apply {
        if (i and 1 == 1) add(Signal.WINK)
        if (i and 2 == 2) add(Signal.DOUBLE_BLINK)
        if (i and 4 == 4) add(Signal.CLOSE_YOUR_EYES)
        if (i and 8 == 8) add(Signal.JUMP)
        if (i and 16 == 16) reverse()
    }

    // works but no functional enough
    fun calculateHandshake2(i: Int): List<Signal> {
        val result = ArrayList<Signal>()
        if (i and 1 == 1) result.add(Signal.WINK)
        if (i and 2 == 2) result.add(Signal.DOUBLE_BLINK)
        if (i and 4 == 4) result.add(Signal.CLOSE_YOUR_EYES)
        if (i and 8 == 8) result.add(Signal.JUMP)
        if (i and 16 == 16) result.reverse()
        return result
    }
}
